src/scipp/scipy/optimize/__init__.py

```diff
@@ -106,18 +173,83 @@ def _parse_bounds(
     return bounds_array[0], bounds_array[1]
 
 
+def _curve_fit(
+    coords,
+    f,
+    da,
+    p0,
+    bounds,
+    map_over,
+    unsafe_numpy_f,
+    out,
+    **kwargs,
+):
+    dg, dgcov = out
+
+    if len(map_over) > 0:
+        dim = map_over[0]
+        for i in range(da.sizes[dim]):
+            _curve_fit(
+                coords,
+                f,
+                da[dim, i],
+                p0,
+                bounds,
+                map_over[1:],
+                unsafe_numpy_f,
+                (dg[dim, i], dgcov[dim, i]),
+                **kwargs,
+            )
+
+        return
+
+    for c in coords:
+        da = _drop_masked(da, da.coords[c].dim)
+
+    X = np.vstack(
+        [
+            np.broadcast_to(
+                da.coords[c].values.reshape(
+                    *(da.sizes[d] if d in da.coords[c].dims else 1 for d in da.dims)
+                ),
+                da.shape,
+            ).ravel()
+            for c in coords
+        ]
+    )
```

Comment:
Check if `da.flatten` (potentially followed by a `fold`) can do this more concisely.